### PR TITLE
Backport min-height block support feature

### PR DESCRIPTION
**Pull request by andrewserong**
This PR backports the minimum height block support feature's PHP changes from Gutenberg. Those changes were originally merged in the following PRs:

* https://github.com/WordPress/gutenberg/pull/45334 and
* https://github.com/WordPress/gutenberg/pull/45300

The opting-in of the block support for particular blocks, and UI support will land separately as part of the main JS package update for 6.2 — hopefully, this PHP change should be able to safely land in isolation.

Trac ticket: https://core.trac.wordpress.org/ticket/57582

Aside from the included tests, I'm not sure if there's a reasonable way to test this prior to the JS packages update landing, since that will also include the block support opt-ins for group and post-content blocks. However, it might be possible to manually update the `block.json` file for the post template block to include `dimensions.minHeight` under supports, and then add the following block markup to a template in the site editor (e.g. the single template):

```
<!-- wp:post-content {"style":{"dimensions":{"minHeight":"80vh"}},"layout":{"type":"constrained"}} /-->
```

On the site-frontend (assuming the site editor allows you to save the above block markup), the post content block should be rendered with `minHeight` set to `80vh`. If this doesn't work in manual testing, however, then we might need to wait until the JS packages have been updated before testing manually.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/theme/wpThemeJson.php [hunk 1]
```diff
@@ -265,6 +268,9 @@ public function test_get_settings_appearance_true_opts_in() {
 					'color'      => array(
 						'link' => true,
 					),
+					'dimensions' => array(
+						'minHeight' => true
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/theme/wpThemeJson.php]:
```suggestion
						'minHeight' => true,
```

Comment (andrewserong) [reply tests/phpunit/tests/theme/wpThemeJson.php]: Thank you! We both noticed that at the same time 😄

File: tests/phpunit/tests/block-supports/dimensions.php [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+		$this->test_block_name = null;
+	}
+
+	public function tear_down() {
+		unregister_block_type( $this->test_block_name );
+		$this->test_block_name = null;
+		parent::tear_down();
+	}
+
+	public function test_dimensions_style_is_applied() {
```
Comment (costdev) [inline tests/phpunit/tests/block-supports/dimensions.php]: These test methods look very similar, differing only on the value of the block name, `supports.dimensions` and of `$expected`.

If that's correct, I'd suggest using a data provider that supplies the `string $block_name`, `array $dimensions` and `array $expected` parameters, which should be documented in the test method's docblock. Then the docblock would look something like:

```
Tests that X does/does not do something.

@ticket 57582

@covers ::wp_apply_dimensions_support

@dataProvider data_name

@param...
```

File: tests/phpunit/tests/block-supports/dimensions.php [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	private $test_block_name;
+
+	public function set_up() {
+		parent::set_up();
+		$this->test_block_name = null;
+	}
+
+	public function tear_down() {
+		unregister_block_type( $this->test_block_name );
+		$this->test_block_name = null;
```
Comment (costdev) [inline tests/phpunit/tests/block-supports/dimensions.php]: A couple of questions:

1. Does the parent `::tear_down()` method not unregister block types?
2. Is it necessary to set new block names for each test, or can we make `$test_block_name` a `static` property and set it in `public static function set_up_before_class()` instead?

**Review by costdev — CHANGES_REQUESTED** [review]
Thanks for the PR @andrewserong! I've left few thoughts below :slightly_smiling_face:

File: src/wp-includes/class-wp-theme-json.php [hunk 4]
```diff
@@ -490,6 +498,7 @@ public static function get_element_class_name( $element ) {
 	 * Options that settings.appearanceTools enables.
 	 *
 	 * @since 6.0.0
+	 * @since 6.2.0 Added `dimensions.minHeight`
```
Comment (mukeshpanchal27) [inline src/wp-includes/class-wp-theme-json.php]:
```suggestion
	 * @since 6.2.0 Added `dimensions.minHeight`.
```

File: src/wp-includes/block-supports/dimensions.php [hunk 5]
```diff
@@ -29,8 +29,7 @@ function wp_register_dimensions_support( $block_type ) {
 		return;
 	}
 
-	$has_dimensions_support = block_has_support( $block_type, array( '__experimentalDimensions' ), false );
-	// Future block supports such as height & width will be added here.
+	$has_dimensions_support = block_has_support( $block_type, array( 'dimensions' ), false );
```
Comment (hellofromtonya) [inline src/wp-includes/block-supports/dimensions.php:32]: Results from a wp.org search:
- 0 plugins found using `__experimentalDimensions` https://wpdirectory.net/search/01GR1JTT957N4BM1FAZAHPCQ1C
- 0 themes found using  `__experimentalDimensions` https://wpdirectory.net/search/01GR1JX1E9HM3BQQQ0GJ457B68

So I'm wondering: 

- What are the backwards-compatibility risks for users who may have a custom plugin or theme using it?
- How was `'__experimentalDimensions'` previously exposed and used?
- What is the mitigation plan for `'__experimentalDimensions'` to upgrade blocks using this setting?

Comment (hellofromtonya) [reply src/wp-includes/block-supports/dimensions.php:32]: One way to mitigate:
- Check it here with an `||`.
- Throw a deprecation notice to alert to use the new `'dimensions'` instead.

File: src/wp-includes/block-supports/dimensions.php [hunk 6]
```diff
@@ -44,23 +43,39 @@ function wp_register_dimensions_support( $block_type ) {
...
+ * @since 6.2.0 Added `minHeight` support.
  * @access private
  *
  * @param WP_Block_Type $block_type       Block Type.
  * @param array         $block_attributes Block attributes.
  * @return array Block dimensions CSS classes and inline styles.
  */
 function wp_apply_dimensions_support( $block_type, $block_attributes ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
-	if ( wp_should_skip_block_supports_serialization( $block_type, '__experimentalDimensions' ) ) {
+	if ( wp_should_skip_block_supports_serialization( $block_type, 'dimensions' ) ) {
```
Comment (hellofromtonya) [inline src/wp-includes/block-supports/dimensions.php:54]: Same BC discussion here. See https://github.com/WordPress/wordpress-develop/pull/3940/files#r1090737459 and https://github.com/WordPress/wordpress-develop/pull/3940/files#r1090745794.

File: src/wp-includes/block-supports/dimensions.php (hunk 5 above)
Comment (andrewserong) [reply src/wp-includes/block-supports/dimensions.php:32]: Thanks for raising this one @hellofromtonya — this is a pretty interesting case for stabilisation in that the support technically existed prior to this PR (with the name `__experimentalDimensions`) however it was a no-op, in that it didn't actually expose any controls or styles. So even if a plugin had opted in to using it (which is pretty unlikely), there would have been no behaviour from it. Because `min-height` is a new feature, and the dimensions controls were never exposed or promoted anywhere, I don't think we need to add any handling to deal with the former `__experimentalDimensions` name — it appears it was mostly there as a placeholder in code for future features.

What do you think?

File: tests/phpunit/tests/block-supports/dimensions.php (hunk 2 above)
Comment (andrewserong) [reply tests/phpunit/tests/block-supports/dimensions.php]: Oh, good idea! That simplifies the tests quite a bit moving over to a data provider, I've pushed an update.

File: tests/phpunit/tests/block-supports/dimensions.php (hunk 3 above)
Comment (andrewserong) [reply tests/phpunit/tests/block-supports/dimensions.php]: Good questions!

1. It doesn't appear to. The teardown methods mostly appear to tear down database related things, so for tests that include registering blocks, it appears that many of the other similar test files register / unregister in setup / teardown methods.
2. I'm not sure if it's strictly necessary to have unique names for each test block, but since it's testing registering a block with unique dimensions support params, I like the idea of it also having a unique block name at registration time, in that we know it won't be re-using an existing test block name. Also, the patterns in this test were copied from the other block supports test files (`border.php`, `colors.php`, `spacing.php`, `typography.php`), and I think it helps readability a bit if we stick to a consistent convention across those files in terms of registering / unregistering blocks with block supports.

So from my perspective, I think the data provider helped simplify the readability of the tests, but personally I'd probably lean toward sticking with the register/unregister convention from the other block supports tests.

File: src/wp-includes/block-supports/dimensions.php (hunk 5 above)
Comment (ntsekouras) [reply src/wp-includes/block-supports/dimensions.php:32]: If it was a no-op, I think we're good to leave it as is.

Comment (hellofromtonya) [reply src/wp-includes/block-supports/dimensions.php:32]: Thank you @andrewserong and @ntsekouras for the explanation. 

tl;dr
No BC concern.

Let me summarize the reasoning:

* Yes, this `__experimentalDimensions` setting existed before 6.2.
* It did nothing. No functionality was attached to it. (think of it as placeholder for the future, ie today)
* If a developer is using it, there is no effect to them for removing it.

I agree then. It's not a BC concern to remove it. Thank you!

File: src/wp-includes/block-supports/dimensions.php (hunk 6 above)
Comment (hellofromtonya) [reply src/wp-includes/block-supports/dimensions.php:54]: tl;dr There's no BC concern as it was never wired to anything before 6.2. See explanation https://github.com/WordPress/wordpress-develop/pull/3940#discussion_r1092546821.

File: tests/phpunit/tests/block-supports/dimensions.php [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+			),
+		);
+
+		$actual = wp_apply_dimensions_support( $block_type, $block_attrs );
+
+		$this->assertSame( $expected, $actual );
+	}
+
+	/**
+	 * Data provider for test_minimum_height_block_support().
```
Comment (costdev) [inline tests/phpunit/tests/block-supports/dimensions.php]:
```suggestion
	 * Data provider.
```
After discussion with other contributors yesterday: For now, data provider docblocks should just be described as "Data provider." [Context](https://github.com/WordPress/wordpress-develop/pull/3920#discussion_r1092021101)

Comment (andrewserong) [reply tests/phpunit/tests/block-supports/dimensions.php]: Good to know, thank you!

**Review by ntsekouras — APPROVED** [review]
Thank you @andrewserong !

I have manually tested the front end by adding
```
<!-- wp:post-content {"style":{"dimensions":{"minHeight":"80vh"}},"layout":{"type":"constrained"}} /-->
```
and adding the support in the block, and works as expected in the front end - that means applying the `minHeight` styles.

**Comment by hellofromtonya** [comment]
Status: Currently reviewing and testing for commit consideration.

File: tests/phpunit/tests/block-supports/dimensions.php [hunk 8]
```diff
@@ -0,0 +1,104 @@
+<?php
```
Comment (hellofromtonya) [inline tests/phpunit/tests/block-supports/dimensions.php]: The file and the class need renaming for Core's Test coding standards and consistency.

* File `wpApplyDimensionsSupport.php`
* Test Class: `Tests_Block_Supports_wpApplyDimensionsSupport`

Will push a commit shortly.

Comment (hellofromtonya) [reply tests/phpunit/tests/block-supports/dimensions.php]: Whoops forgot to share the naming convention:

For the file:
It's the function's name in camelCase without the `_`.

For the test:
It's in this format:
```
Tests_{group}_{functionNameInCamelCase}
```

See handbook here https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#test-classes for more information.

Note: In Gutenberg, more work is needed in the linting before its tests can be converted into Core's coding standards.

**Review by hellofromtonya — APPROVED** [review]
* Confirmed the PHP changes from the GB PRs are here ✅ 
* Confirmed there's no BC issue for the removal of `__experimentalDimensions` setting ✅ 
* Reviewed for coding standards ✅ 
* Tests are okay ✅ 
* @ntsekouras tested it ✅ 

LGTM 👍  Will prepare the commit once the CI jobs finish and are 🟢

File: tests/phpunit/tests/block-supports/dimensions.php (hunk 8 above)
Comment (costdev) [reply tests/phpunit/tests/block-supports/dimensions.php]: @hellofromtonya According to the handbook, it says the test class should be in this format:
```
Tests_{Group}_{FunctionNameInTitleCase}
```

> Thus, the class `Tests_Comment_GetCommentClass()`, which contains tests for the `get_comment_class()` function, is located in `tests/comment/getCommentClass.php`.

What I'm not sure about though, is whether `{Group}` should be `BlockSupports`, or `Block_Supports`, i.e.

- `Tests_BlockSupports_WpApplyDimensionsSupport` OR
- `Tests_Block_Supports_WpApplyDimensionsSupport`

Comment (hellofromtonya) [reply tests/phpunit/tests/block-supports/dimensions.php]: @costdev The other test in the group has it as `Block_Supports`. So I felt it that way for consistency and because the handbook is unclear if the group should also be smooched together. Thinking that topic needs a Trac ticket of its own to discuss and get consensus.

Comment (costdev) [reply tests/phpunit/tests/block-supports/dimensions.php]: Regarding `_Wp` rather than `_wp`, see [this ticket](https://core.trac.wordpress.org/ticket/56846) with support for encouraging `Wp` for consistency among all test classes, despite there being more cases of `_wp` in the test suite historically. I'm happy to go with what you decide regarding the capitalization after viewing the ticket. 🙂

Comment (hellofromtonya) [reply tests/phpunit/tests/block-supports/dimensions.php]: Oh wait I do see what you mean though about capitalizing the first letter of the function's name. Yup that needs changing. Good catch. Will do that in the commit.

Comment (hellofromtonya) [reply tests/phpunit/tests/block-supports/dimensions.php]: Capitalization done here https://github.com/WordPress/wordpress-develop/pull/3940/commits/12cebb4cd4239b4d9c06a05c20ccd2c21b5f55fb. Thanks @costdev!

**Comment by hellofromtonya** [comment]
Committed via https://core.trac.wordpress.org/changeset/55175.

**Comment by andrewserong** [comment]
Wonderful, thank you for landing this one, and for the extra discussion, I learn more with each of these PRs 🙂
